batch-processing/main.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"context"
+	"encoding/csv"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"gopkg.in/gographics/imagick.v2/imagick"
+)
+
+func main() {
+	// We need a file to read from
+	input := flag.String("input", "", "A path to a CSV with a `url` column, containing URLs for images to be processed")
+	flag.Parse()
+	if *input == "" {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	awsRoleUrn := os.Getenv("AWS_ROLE_URN")
+	if awsRoleUrn == "" {
+		log.Fatalln("Please set AWS_ROLE_URN environment variable")
+	}
+	s3Bucket := os.Getenv("S3_BUCKET")
+	if awsRoleUrn == "" {
+		log.Fatalln("Please set S3_BUCKET environment variable")
+	}
+
+	// Set up S3 session
+	// All clients require a Session. The Session provides the client with
+	// shared configuration such as region, endpoint, and credentials.
+	sess := session.Must(session.NewSession())
+
+	// Create the credentials from AssumeRoleProvider to assume the role
+	// referenced by the ARN.
+	creds := stscreds.NewCredentials(sess, awsRoleUrn)
+
+	// Create service client value configured for credentials
+	// from assumed role.
+	svc := s3.New(sess, &aws.Config{Credentials: creds})
+
+	// Create a context with a timeout that will abort the upload if it takes
+	// more than the passed in timeout.
+	ctx := context.Background()
+
+	// Set up imagemagick
+	imagick.Initialize()
+	defer imagick.Terminate()
+
+	// Open the file supplied
+	in, err := os.Open(*input)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Read the file using the encoding/csv package
+	r := csv.NewReader(in)
+	records, err := r.ReadAll()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// TODO: validate `records`
+
+	for i, row := range records[1:] {
+		url := row[0]
+
+		inputFilepath := fmt.Sprintf("/tmp/%d-%d.%s", time.Now().UnixMilli(), rand.Int(), "jpg")
+		outputFilepath := fmt.Sprintf("/tmp/%d-%d.%s", time.Now().UnixMilli(), rand.Int(), "jpg")
+
+		log.Printf("downloading: row %d (%q) to %q\n", i, url, inputFilepath)
+
+		// Create a new file that we will write to
+		inputFile, err := os.Create(inputFilepath)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer inputFile.Close()
+
+		// Get it from the internet!
+		res, err := http.Get(url)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer res.Body.Close()
+
+		// Ensure we got success from the server
+		if res.StatusCode != http.StatusOK {
```

Comment:
As an extension problem in the README, I'd add trying to get them thinking about pause/resume, so that already-done work can be skipped if we end up re-running (i.e. if the processed failed, we somehow have tracked what we've already processed, so we can just pick up the not-yet-done work). But not for the first cut!

However, for the example impl I'd probably add a comment here contrasting the pros of eagerly failing vs log and continue